src/OrasProject.Oras/Registry/Remote/Repository.cs

```diff
@@ -371,6 +397,195 @@ internal Reference ParseReferenceFromContentReference(string reference)
     public async Task MountAsync(Descriptor descriptor, string fromRepository, Func<CancellationToken, Task<Stream>>? getContent = null, CancellationToken cancellationToken = default) 
         => await ((IMounter)Blobs).MountAsync(descriptor, fromRepository, getContent, cancellationToken).ConfigureAwait(false);
 
+    /// <summary>
+    /// ReferrersAsync retrieves referrers for the given descriptor and artifact type if specified
+    /// and return a streaming of descriptors asynchronously for consumption.
+    /// If referrers API is not supported, the function falls back to a tag schema for retrieving referrers.
+    /// If the referrers are supported via an API, the state is updated accordingly.
+    /// </summary>
+    /// <param name="descriptor"></param>
+    /// <param name="artifactType"></param>
+    /// <param name="cancellationToken"></param>
+    public async IAsyncEnumerable<Descriptor> ReferrersAsync(Descriptor descriptor, string? artifactType,
+        [EnumeratorCancellation] CancellationToken cancellationToken = default)
+    {
+        if (ReferrersState == Referrers.ReferrersState.NotSupported)
+        {
+            // fall back to tag schema to retrieve referrers
+            await foreach (var referrer in ReferrersByTagSchema(descriptor, artifactType, cancellationToken)
+                               .ConfigureAwait(false))
+            {
+                yield return referrer;
+            }
+
+            yield break;
+        }
+        
+        await foreach (var referrer in ReferrersByApi(descriptor, artifactType, cancellationToken)
+                           .ConfigureAwait(false))
+        {
+            // If Referrers API is supported, then it would return referrers continuously
+            // otherwise, this line of code is not executed
+            // and the ReferrerState would be set to false in the method ReferrersByApi.
+            yield return referrer;
+        }
+        
+        if (ReferrersState == Referrers.ReferrersState.NotSupported)
+        {
+            // fall back to tag schema to retrieve referrers
+            await foreach (var referrer in ReferrersByTagSchema(descriptor, artifactType, cancellationToken)
+                               .ConfigureAwait(false))
+            {
+                yield return referrer;
+            }
+        }
+    }
```

Comment:
nit: We may add a few more comments to make it clearer, e.g.
```suggestion
        // referrers state is unknown or supported
        await foreach (var referrer in ReferrersByApi(descriptor, artifactType, cancellationToken)
                           .ConfigureAwait(false))
        {
            // If Referrers API is supported, then it would return referrers continuously
            // otherwise, this line of code is not executed
            // and the ReferrerState would be set to false in the method ReferrersByApi.
            yield return referrer;
        }
        
        if (ReferrersState == Referrers.ReferrersState.NotSupported)
        {
            // referrers state is set to NotSupported by ReferrersByApi, fall back to tag schema to retrieve referrers
            await foreach (var referrer in ReferrersByTagSchema(descriptor, artifactType, cancellationToken)
                               .ConfigureAwait(false))
            {
                yield return referrer;
            }
        }
    }
```